src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,179 @@
+    private Node<T> tail;
+    private int size;
+
+    static class Node<T> {
+        private T element;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(Node<T> prev, T element, Node<T> next) {
+            this.prev = prev;
+            this.element = element;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (isEmpty()) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        isIndexHigherOrLowerThanSize(index);
+        if (isEmpty()) {
+            head = tail = new Node<>(null, value, null);
+        } else if (index == 0) {
+            addFirstElement(value);
+        } else if (index == size) {
+            addLastElement(value);
+        } else {
+            addMiddleElement(value, index);
+        }
+        size++;
     }
 
     @Override
-    public void addAll(List<T> list) {
+    public T set(T value, int index) {
+        isIndexValid(index);
+        Node<T> nodeToUpdate = findNodeByIndex(index);
+        T oldValue = nodeToUpdate.element;
+        nodeToUpdate.element = value;
+        return oldValue;
     }
 
     @Override
-    public T get(int index) {
-        return null;
+    public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
-    public T set(T value, int index) {
-        return null;
+    public T get(int index) {
+        isIndexValid(index);
+        Node<T> foundNode = findNodeByIndex(index);
+        return foundNode.element;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isIndexValid(index);
+        Node<T> nodeToRemove = null;
+
+        if (index == 0) {
+            nodeToRemove = removeFirstNode();
+        } else if (index == size - 1) {
+            nodeToRemove = removeLastNode();
+        } else {
+            nodeToRemove = removeNodeAtIndex(index);
+        }
+
+        size--;
+        return nodeToRemove.element;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        int counter = 0;
+        while (currentNode != null) {
+            if (currentNode.element == object || (currentNode.element != null
+                    && currentNode.element.equals(object))) {
+                remove(counter);
+                return true;
+            }
+            currentNode = currentNode.next;
+            counter++;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
```

Comment:
[Don’t iterate your nodes from the head to tail or vice versa.](https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html#dont-iterate-your-nodes-from-the-head-to-tail-or-vice-versa)